### ISSUE TITLE
OdeSystem to NonlinearSystem to OptimizationSystem

### ISSUE THREAD

**Issue by ValentinKaisermayer** [SciML/ModelingToolkit.jl#1916]
Use-case: nonlinear model predictive control (MPC)

MTK used to have this a while ago.

**Comment by wangyunyueyueniao** [comment]
Agree, and that also make steady-state and dynamic simulation have a shared code base.
[link to discourse](https://discourse.julialang.org/t/how-to-convert-an-odesystem-to-nonlinearsystem/91079?u=wangyunyueyueniao)
